### PR TITLE
news feature requirements implementation

### PR DESCRIPTION
**Pull request by rajavarra**
**Endpoints**

- **POST /register:** Register a new user.
- **POST /login:** Log in a user.
- **GET /news:** Fetch news articles based on the logged-in user's preferences.
- **PUT /news/preferences:** Update the news preferences for the logged-in user.
- **GET /news/preferences:** Retrieve the news preferences for the logged-in user.

**optional extensions will  be implemented soon**

### CODE REVIEW

File: src/controllers/newsController.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+const newsService = require('../services/newsService');
+const getAllNews = async (req, res) => {
+  try {
+    const { user } = req;
+    const news = await newsService.getAllNews(user);
+    res.status(200).send(news);
+  } catch (error) {
+    res
+      .status(error?.status || 500)
+      .send(error?.message || 'Internal Server Error.');
```
Comment (suyash-thakur) [inline src/controllers/newsController.js:10]: Try not to expose the server error on client side

File: src/controllers/userController.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+const bcrypt = require('bcrypt');
+const userService = require('../services/userService');
+
+const signUp = (req, res) => {
+  const { fullname, email, password } = req.body;
+
+  const newUser = {
+    id: uuid(),
+    fullname,
+    email,
```
Comment (suyash-thakur) [inline src/controllers/userController.js]: use toLowerCase

File: src/controllers/userController.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    password: bcrypt.hashSync(password, 8),
```
Comment (suyash-thakur) [inline src/controllers/userController.js]: Avoid dangling variables

File: src/services/newsService.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  return response.data.articles;
+};
+const getPreferences = (reqUser) => {
+  const user = userService.users.find((user) => user.email === reqUser.email);
+  const preferences = user.preferences;
+  return preferences;
+};
+const modifyPreferences = (reqUser, userPreferences) => {
+  const user = userService.users.find((user) => user.email === reqUser.email);
+  user.preferences = [...user.preferences, ...userPreferences.preferences];
```
Comment (suyash-thakur) [inline src/services/newsService.js]: Check for duplicate prefrences

File: src/services/userService.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+  if (findUserIndex === -1) {
+    const error = new Error(`User with email ${email} does not exist`);
+    error.status = 400;
+    throw error;
+  }
+
+  const match = await bcrypt.compare(password, users[findUserIndex].password);
+
+  if (match) {
```
Comment (suyash-thakur) [inline src/services/userService.js]: if(!match) { ...  return }
....

File: src/controllers/userController.js (hunk 2 above)
Comment (rajavarra) [reply src/controllers/userController.js]: resolved this in new commit

File: src/controllers/userController.js (hunk 3 above)
Comment (rajavarra) [reply src/controllers/userController.js]: handled with config file

File: src/services/newsService.js (hunk 4 above)
Comment (rajavarra) [reply src/services/newsService.js]: right now as fix, overriding the existing list with new
in upcoming iteration , will be using predefined preferences with flags

File: src/services/userService.js (hunk 5 above)
Comment (rajavarra) [reply src/services/userService.js]: changed as per comment

File: src/controllers/newsController.js (hunk 1 above)
Comment (rajavarra) [reply src/controllers/newsController.js:10]: used common error handler for users Controller, will be working on more centralised error handling in v2

**Pull request merged**